### ISSUE TITLE
AniList API v2

### ISSUE THREAD

**Issue by joshstar** [joshstar/AniList-API-Docs#63]
Currently API v1 is fine but has a large number of issues and inconsistencies, this issue is to discuss possible API breaking improvements and features for a new second version of the API.

Please submit any ideas or changes you would like to see for API v2 here.
I would especially love to hear from developers who have used our API in production (@blitzprog, @Kazakuri, @z411, @ratan12, @Nihilate) and developers that have worked on or with other public APIs.

Time frame for API v2: (This is just an initial idea, very much likely to change) 
- Q3 2016 refactor AniList backend and start initial API v2 work
- Q3 2016 AniList new backend released
- Q3/4 2016 release alpha version of API v2 to public, work with community to stable
- Q4/Early Q1 2017 release stable API v2
- +6 months after API v2 release, discontinue API v1 

Here are a few initial changes I would like to see:
- **Consistent types**, at the moment in API v1 some int are strings, and some booleans are ints (0 or 1). In API v2 everything should be cast to their correct type before output, this can be done easily with Laravel 5's cast feature.
- **More Flexibility**, API v1 is too restrictive with small data models and relational data, API v2 should provide more data and options, possibly even providing [SQLish functionality to the API endpoints](https://github.com/marcelgwerder/laravel-api-handler#filtering).
- **Consistent return models**, some endpoints in API v1 return data along with pagination data, others don't for no real reason. Some errors are returned in a JSON format, while others there is no JSON output and instead developers much rely on HTTP status code. All of this should be standardized for every API output.
- **More Meta Data**, (this continues on for the previous point) API v2 should provide a meta data object with every model, with information like counts, current API url, model API url, model AniList.co site url etc.

**Comment by blitzprog** [comment]
* **Consistent types**: Nice, I noticed this querying browse/anime. Some IDs are strings, some are ints. Not a big deal post-converting it to int but it's good that this will be standardized.
* **Consistent return models**: Good point. Get rid of the **small anime model**. Either return a) full info or b) only the ID. Something inbetween is like you can't decide between two girls who you want as a girlfriend. Once decided, be consistent.
* **Authorization** by sending the token in the headers never worked for me. I always need to send it as a query URL parameter. On the contrary, Hummingbird v1 API never complained about sending the token as a header. Maybe I'm just too stupid *shrugs*
* Not really an API feature but I'd like to be able to retrieve **cover images** bigger than *reg* size. Not sure if this is possible but Hummingbird has some really big cover images.
* **Specify precisely** how much time should pass between 2 sub-sequent requests when doing let's say 500 queries (for 500 users, e.g.). I ended up using a 1100 ms interval to reduce the load on Anilist and in order to not get blocked by automatic Cloudflare DoS protection. It would be nice if we knew exactly which interval is the bare minimum. However you shouldn't manually enforce it because we also need "bursts" of requests sometimes (e.g. in [notify.moe](https://github.com/animenotifier/notify.moe) the Chrome extension allows you to force-reload your anilist data by clicking on the button which results in multiple requests in <1 second).
* I'd like to see an endpoint `/api/anime/all` which does nothing but return an array of **all anime IDs** on anilist. I believe it'd be an array of around 10k integers and this isn't going to be queried a lot (if app developers use it properly).
* If the above point of getting IDs is implemented, I'd like to see [Anime News Network Encyclopedia API](http://www.animenewsnetwork.com/encyclopedia/api.php) like **batching**. Something like `/api/anime/1,12,7,9` (you can use another separator character or format, it's just for explaining) would return an array of 4 full anime models. Limit it to 50 IDs per batch like they do.

All I can think of at the moment. Glad to see this is getting a new version :+1:

**Comment by Josh-Star** [comment]
- Small data models might still be available but probably just as an option for when working with clients running over mobile data for example.
- Not sure why the Authorization header doesn't work for you, but I'll look into it.
- Our regular size cover images are the largest we have at the moment, we'll be looking at sourcing some larger when once we get some more hard drive capacity as we're running a little low at the moment.
- Request limits is entirely a cloudflare thing I have no control over really, I need to do some work on the back end before we'll be able to remove cloudflare but that should be sorted way before API v2 is out. But documentation is also something I want to do much better and more of for v2.
- I don't see any reason why we couldn't offer an end point with all the anime ids but I do have to wonder why anyone would really need that?
- Batching and more advanced queries will definitely be available, I already mentioned this in the main post but [see here for some of the other kind of things that might be possible](https://github.com/marcelgwerder/laravel-api-handler#filtering).

**Comment by blitzprog** [comment]
`/api/anime/all` would make coding the synchronization with anilist a really nice one-liner:

```js
req('/api/anime/all').then(all => all.map(anime => req('/api/anime/' + anime.id)))
```

Beautiful code if you ask me (and req would be set to 1 sec query interval or batches requests in v2).

**Comment by Josh-Star** [comment]
I'm really not following why you would ever need to do that, you would just be scraping the entire anime table? I don't think that's really making appropriate use of the api.

**Comment by ratan12** [comment]
- The headers are working fine here.
@blitzprog Make sure that you send this:
header: "Authorization"
value: "Bearer " followed by the access_token
- It would be nice if you add a test page in the settings -> apps. With test page I mean a page where we could enter the url like: user/ratan12/animelist which will return the proper json format trough the API. So no auth or something. This will actually save time when developing and will only be available if the developer is logged in AL
- How are you planning to version the API? "https://anilist.co/api2/" or "https://anilist.co/api/2/"
- It will be nice to mark new changes in the documentation.
- Make sure that users can use the API on multipe devices
- Crunchyroll url would've be awesome

**Comment by Josh-Star** [comment]
- Versioning might be via a header, or possibly /api2/ like you suggested, its something we can decide on at a later date.
- Crunchyroll url is already avaliable in api v1 in the external links array

**Comment by blitzprog** [comment]
@ratan12: Will try. I might have used a different key.
@Josh-Star: My use case might be special since I need a local copy for latency / performance reasons. Especially since you said that some titles (see latest twitter) are not gonna be included on anilist.

- Add my own titles manually (which don't exist on anilist)
- Performance
- Latency

**Comment by Josh-Star** [comment]
@blitzprog If you absolutely need a local copy for latency/performance reasons I seriously suggest you reconsider how your application works. I'm afraid thats simply not something I am willing to support as it almost defeats the entire purpose of a public API.

**Comment by blitzprog** [comment]
@Josh-Star: Yes it's absolutely required. The performance when permanently querying is utterly unacceptable. Another nice side effect from your perspective is that it actually decreases the load on the anilist server. We actually discussed this on twitter already (the story about unifying provider titles and IDs).

You might simply not understand the background of this decision. I agree about not spamming the thread about my special case, I'll send you a message on Twitter about this.

**Comment by soruly** [comment]
I agree with blitzprog.
Main reasons why I store a local copy of anilist data (for the moment).
* Localization
* Complex Queries (with elasticsearch)
* Latency (Search as you type)
* Add my own titiles

I am translating anime info to chinese (since there is nothing like anilist websites in Hong Kong / Taiwan). As you see in my search-anime-by-screenshot demo site, I have added my own chinese title and alias translated by myself. Since small models does not include staff and studio, now I have to crawl all anime small models from browse/anime (from page > 230), extract the IDs, then use the IDs to query one by one for full anime model.

Internally I have a set of merged data loaded in elasticsearch. The fuzzy search and prefix search is particularly useful when I forgot how to spell the title. I can also search animes by either Japanese / Romanji / English / Chinese title or alias. I can even search with complex queries with filter and sort results by rating / popularity / drop rate. It would be nice if APIv2 supports complex queries like that. But it would be hard to provide a comprehensive functionality like elasticsearch.

The API's response time is fast, but not fast enough for search-as-you-type, especially I am on the other side on earth.

As blitzprog mentioned, I would like to add some animes that are not in anilist such as doujin animes like 幻想万華鏡 or 艦ほり！

I understands that the API is mainly for users of anilist, so it's up to you to decide what functions that should / should not be provided. There must be some personal needs that anilist can't provide, therefore we "Fork" part of anilist and customize our needs. But whenever there is forking, there is also merging. It would be great to see an API not just fetching data from anilist, but an API of exchanging / aggregating various sources of anime data.

I have a few use cases of the API that I would like to share with, but let's discuss this on twitter and keep this discussion technical.

**Comment by Nihilate** [comment]
A little late to the party, but for [my current needs](https://www.reddit.com/user/Roboragi) there aren't that many things I'd really want added to V2.

The biggest issue I face is making sure that the results I get are accurate for what people are requesting, and the largest part of that is standardising results across multiple databases (9 times out of 10 the problem is the MAL API giving me gibberish XML). It's mostly a data issue, but being able to have SQL-like functionality to the end points would let me provide finer control to users.

Beyond that, being able to search/access the large anime model more easily would be great and having an endpoint for LNs totally separate from manga (or at least an easy way to filter for them) would be cool too.

One thing that would be really cool to have access to is anonymous user data - something that would let me provide recommendations based on relationships between shows (e.g. users who rated X highly also rated Y highly).

**Comment by blitzprog** [comment]
> Crunchyroll url is already avaliable in api v1 in the external links array

I just checked. It's not. [API](https://anilist.co/api/anime/21234) (add access token to the link) vs. [Page](https://anilist.co/anime/21234). I see no external links array, or is it obtainable somewhere else?

**Comment by danejur** [comment]
Hey Josh, I'm not sure if this should go here or on the forums, so let me know if should be moved over there.

A feature that I feel is currently lacking in the API is the concept of "similar anime/manga". By that, I don't mean items that are related to other items (sequels, alternate telling, etc.), but things that are "similar" (e.g. Gurren Lagann is similar to Kill la Kill). That example might make sense to some and not to others, so this feature would be user moderated. For instance, as a user, I would add a "show similarity" on TTGL to KLK (maybe selecting from a list of reasons why, how similar, etc.) At that point, nothing would appear to the end user, but other users who wish to also suggest similar shows can essentially "vote" on that one (by themselves selecting it as similar or not). After a certain number of users votes on it, if the percentage that agree with it being similar is higher than some threshold (say 75%), it will appear publicly as a "similar anime".

AniDB has had this system for a while, and it is one of the very few things I've missed from their API since switching to yours (I assure you, yours is leaps and bounds better). Let me know if you consider this a good idea, or if it goes against the kind of data that you want to present to the user on the site.

**Comment by joshstar** [comment]
@danejur Sorry for the crazy late reply, thats not really related to APIv2 but is definitely something we've been wanting to add for a long time, at this point it will be something we work on after APIv2 is out (or in beta) so you can expect to see it in v2 at some point in the future.

**Comment by danejur** [comment]
@joshstar Thanks for the response! I imagined that might be the case. Looking forward to it.

Another small (possibly large) feature I was considering asking about was some sort of "hash" or "last updated timestamp" field. I'd like to locally cache whatever static data I can (such as anime/manga/staff info), but I'd definitely like to have a way to determine if any changes have been made to the object before pulling the entire thing. I'd rather hit something like /anime/{id}/hash or /anime/{id}/lastupdated to get a simple hash or datetime back before fetching the entire object (and hopefully saving the server some unneeded cycles).

If this seems like a stupid idea to you, please disregard. Also, if you'd rather us not cache anything, that is fine, too. Your call, of course!

**Comment by joshstar** [comment]
@danejur We'll be exposing our model update times for v2, and then you should be able to get the latest updated anime/manga via something like anime?sort=updated_time. Our update/created times are stored on the model so I'm not sure if we would save much by just returning the timestamp but I don't see any reason why we couldn't have that as an optional route.

**Comment by danejur** [comment]
@joshstar Cool. If anything, it could save bandwidth, but just including the time stamp as a part of the returned model would be cool, as well. Thanks again!

**Comment by joshstar** [comment]
Interested to here all of you're opinions on this: integer or string typed values for api v2.

Currently for values like anime status (airing, finished) and list status (watching, completed) we provide and expect string values, however an alternative for v2 would be to just provide/expect the actual ints that are stored in the database for these values (0, 1) (0 => airing, 1 => finished).

Either way would be fine for me really but I'm interested in which you would prefer to use in your applications. I guess strings would be easier for UIs, and make code a bit more human readable but they can sometimes be a bit more of a pain to deal with.

**Comment by soruly** [comment]
@joshstar I would prefer using integers. It's good for mapping the status to various language. To check airing status, we should use enum instead of comparing string
```
const AiringStatus = {
  AIRING: 0,
  FINISHED: 1
};
if(anime.airing_status == AiringStatus.AIRING){ ... }
```
`if(anime.airing_status == "currently airing")` Doesn't feel right.

**Comment by blitzprog** [comment]
> I guess strings would be easier for UIs

Objectively speaking this is not a good argument for strings. iirc the strings are not capitalized and UIs require different types of capitalization or even **wording**. In most cases we need to use our own strings **anyway**, every app has different UI & Feel.

Integers have the single (imo minor) disadvantage that it introduces another **learning barrier**.

The advantage of integers is that is a more **DRY approach** and costs (even if not significantly) **less bandwidth**.

I don't really care but :+1: for integers.

**Comment by danejur** [comment]
I agree with @blitzprog and @soruly.

I wouldn't have a problem with shifting most of the predefined values to integers. My main concern would be the method in which we might be informed about updates to constants, such as adding a new status, removing one, or changing one.

**Comment by lolsheeplol** [comment]
Could we have an update on the development status of API v2?  I'm digging most of the changes mentioned above.

**Comment by joshstar** [comment]
Outside of some initial prototyping, I haven't really started work on v2 yet. I've been working on a backend/database refactor along with adding a host of new features and improvements to the site in general. As always these things take a lot longer then initially thought so I've updated the release timeframe in the top post here.

Once this new update is out (sometime in august) I'll start work on v2 and release some early indev releases. Sorry for the delays as usual.

**Comment by AndyIbanez** [comment]
I found another inconsistency issue.

![Some null some empty](https://spiky.io/36.png)

This is happening on an account I created just for testing and I haven't played with any of its settings yet.

I think the best way to approach this would be to return an empty array when there's no custom lists, and when there are, just return the ones that exist with a non-empty name.

**Comment by AndyIbanez** [comment]
Another thing that should be changed for v2 (or sooner).

The user currently has the ability to select what language to show titles in, be it Japanese, Japanese in Romaji, or English.

In Anime and Small User models (presumably their Manga counterparts as well), they return `title_romaji`, `title_japanese`, and `title_english`. I think it would make sense to return just  a`title` field with the title in the  language the user has configured in their settings. While we can of course grab the title system for each user from their `user` page, this will often require an extra request to the API to get that setting.

**Comment by danejur** [comment]
I have to disagree with @AndyIbanez on the matter of returning titles. Choosing which title to display is logic that should be offloaded to the client, not the server. Besides, I'd imagine we want the API to have as much parity as possible with what is presented on the site, and removing the alternate titles would require additional requests to be made.

Can't the user's configuration be cached locally?

**Comment by AndyIbanez** [comment]
I am fine with the API returning all possible titles, but I disagree with the notion that selecting the correct title should be offloaded to the client. Any setting that can be done on Anilist should be respected by the API itself. It's kind of like having the "adult" switch and then having us use the user settings to filter all the adult stuff out when it is off - The API should not return any adult content if the user chose so in the service. Adding some sort of `default_title` field to reflect the title language the user has chosen alongside `title_romaji`, `title_japanese`, and `title_japanese` would be the best solution IMO. Then you can use them a-la-MAL if you fancy.

Settings can be cached locally, but then there's the matter of refreshing those settings. Should I refresh them every hour? On every app launch? If said settings can be changed locally and are (don't think this possible atm, but correct me if I'm wrong), it would be easy to keep the settings locally.

**Comment by AndyIbanez** [comment]
I have noticed that many datetimes return as a month and a date, like this:

```
2016-06-03 09:42:33
```

The latest API calls I have been doing for my wrapper (anime lists and the like) return ISO 8601 strings like this:

```
2016-08-03T02:57:51+09:00
```

Every date property should return the ISO 8601 string to make it easier to parse and represent in the user's current timezone.

I wrote a function to convert the first string into a ISO 8601, based on the information I got from other questions I asked (I really needed to know the server's location or the UTC offset - this was before I noticed other calls do return the offset) to calculate the time in the user's current timezone, but this should not be the developer's responsibility.

**Comment by joshstar** [comment]
Just to give you all a bit of an update on v2, I've been doing a lot of research into how to handle the api queries.

Initially I was looking into creating some kind of url-based query builder that would look something like this: 
`http://anilist.co/api/v2/anime?id-in=1,21,31&with=rankings,relations:limit(2:4).airing&limit=(2:4)` 

This would be crazy flexible but no one would really want to hand write quires like this so the idea would be to create a bunch of open source clients for different platforms/languages to make writing these queries easier, something like this:
```
AlApi->getAnime
     ->whereIn('id', [1, 21, 31])
     ->with('rankings', 'relations' => function($query) { 
          $query->with('airing')->limit(2, 4); 
      })
     ->limit(2, 4);
```

This would be fine but creating our own clients for this would take a fair bit of time and take resources away from improving the site/api.

So recently I've been looking into [GraphQL](http://graphql.org/) more. I recommend looking at site examples to get an idea of how it works, they also have a [site where you can test queries against a star wars api](http://graphql-swapi.parseapp.com/). 
Basically it would be more flexible then the previous method, officially supported by a big company/open source groups, and has a bunch of well tested clients already made for almost every language.

If we do go with GraphQL I would want to supply a few simple non-graphql routes for commonly used models that would then route to GraphQL, just to make really simple apps not require much GraphQL  knowledge and code dependencies.

Interested to hear all your opinions on these, I'm definitely leaning more towards GraphQL at this point as it would be a fair bit easier/quicker for us to develop the api with.

**Comment by AndyIbanez** [comment]
GraphQL looks GREAT. Not only can it help you implement the API but the control it gives to the consumers of the API looks great. I specially like how you can choose what to include and what to exclude and what to include, which has been my main beef with some endpoints up until now.

**Comment by danejur** [comment]
I'm a pretty big proponent of GraphQL, so I would certainly love to see it used here. I think having some static endpoints that return commonly fetched data would still be a good idea, but letting the consumer decide what and how to consume is probably the best way to design an API, in my opinion. +1 to GraphQL.

**Comment by ALawliet** [comment]
+1 GraphQL! Rock on! I'm currently building an anime-related app with GraphQL/React Native, and using your API for anime and character search. Best one out of there out all the ones I've tried. Glad to know you guys are active and that it's going to be future-proof, especially with GraphQL!

**Comment by ratan12** [comment]
It looks promising but I will have to figure out how to handle.

**Comment by wax911** [comment]
I just played around with GraphQL and I like it. I also read through all the comments here, and wow there is a lot of suggestions which I think might not be relevant to everyone, I think we must consider of the general purpose of the API when we give our two sense otherwise everyone will have a request custom tailored to their needs. I'm just saying so please don't shoot me 🎱 

Josh would it be a bad idea to return most of the text descriptions in HTML in the next api release? The anime and manga descriptions already do that but the reviews and user descriptions are formatted to the AniList input fields  e.g. ___Text Here.___ could be returned in the end point as <b>Text Here.</b> and img1(http://link) could be returned as a normal image tag, the break lines can remain as \n not really a problem. I'm not sure if this is relevant or worth bothering with though

**Comment by joshstar** [comment]
Yeah I would prefer to add least offer the option of html instead of markdown. We have a fair bit of custom markdown which makes it a pain for api devs to work with, but it also means that if we want to return it as html we'll need to support that custom markdown on the backend too, but thats probably something we can deal with.

Started work on APIv2 now with graphql, maybe expect a in-dev preview some time soon. Uni and other AniList work is taking a bit of a priority at the moment so I wouldn't expect a stable v2 any time soon, but hopefully something that you can all try out and provide feedback for sooner rather than later.

**Comment by wax911** [comment]
Hey Josh, another suggestion 👍 @joshstar, I've posted something similar on the website, sorry for the repetition but thought it would also avoid another similar post from being created here. I just noticed something strange with new users model, which returns empty arrays for:

`stats -> status_distribution & probably stats -> score_distribution` 

__Resulting in an error such as__:

` Expected BEGIN_OBJECT but was BEGIN_ARRAY at line X column XXX path $.stats.status_distribution.anime` (Java Specific, but may also apply to other languages like OOP languages which are not loosely or dynamically typed)

I understand that the object creation from the server will automatically generate array JSON objects for empty result sets, so I wonder if there is anyway to work around this on your end and return empty object types to assure safe type conversions on clients?

__New User Data Stats Object:__
`
  "stats": {
    "status_distribution": {
      "anime": [],
      "manga": []
    },
    "score_distribution": {
      "anime": [],
      "manga": []
    }
}`

__VS Existing User With Some Anime or Manga Collections:__
`  "stats": {
    "status_distribution": {
      "anime": {
        "watching": 7,
        "plan to watch": 83,
        "completed": 145,
        "dropped": 5,
        "on-hold": 14
      },
      "manga": {
        "reading": 1,
        "plan to read": 16,
        "completed": 0,
        "dropped": 0,
        "on-hold": 0
      }
    },
    "score_distribution": {
      "anime": {
        "10": 0,
        "20": 0,
        "30": 0,
        "40": 1,
        "50": 16,
        "60": 46,
        "70": 46,
        "80": 26,
        "90": 21,
        "100": 0
      },
      "manga": {
        "10": 0,
        "20": 0,
        "30": 0,
        "40": 0,
        "50": 0,
        "60": 0,
        "70": 1,
        "80": 0,
        "90": 0,
        "100": 0
      }
    },
    "favourite_genres": {
      "Action": 100,
      "Comedy": 84,
      "Adventure": 67,
      "Supernatural": 58,
      "Fantasy": 48,
      "Sci-Fi": 34,
      "Romance": 31,
      "Ecchi": 30,
      "Drama": 22,
      "Mystery": 16
    }
  }`

As you can see the status_distribution & score_distribution inner types are objects and not arrays. Can I suggest not providing the status_distribution or score_distribution when there is not data present as you do with the favourite_genres object which isn't shown until the data is available?

**Comment by joshstar** [comment]
@wax911 yeah this is an issue with php's associative arrays being converted to json, if the array has key/values it will be made into an object but if its empty it will just be seen as a normal json array. 

I already have a plan for how to fix this for api v2, haven't gotten around to trying it out yet but pretty sure it will work. 
With the way graphql works I don't think not returning the key/values at all would be a good idea so it will probably just return an empty value of the correct type.

**Comment by wax911** [comment]
Hey josh, I was wondering would it be possible (If you do plan on integrating anything like this) to make a logout API endpoint for V2 third party apps such as AniTrend which would allow something like POST: apiv2/logout body {token_code:XXXXXX} which will remove the authenticated code from the users profile? Not really a must and also don't know if it would be useful for others

**Comment by joshstar** [comment]
AniList Api v2 is now in pre-release 😃 

More info and discussion [here](https://github.com/AniList/ApiV2-GraphQL-Docs/issues/2)
